app-ios/tutanota/Sources/Alarms/AlarmManager.swift

```diff
@@ -144,69 +43,76 @@ class AlarmManager {
     }
     
     TUTSLog("Finished processing \(alarms.count) alarms")
-    self.userPreference.store(alarms: savedNotifications)
+    self.alarmPersistor.store(alarms: savedNotifications)
+    
+    self.rescheduleAlarms()
+    TUTSLog("Finished re-scheduling")
     if let error = resultError {
       throw error
     }
   }
-  
-  private func hasNotificationTTLExpired() -> Bool {
-    guard let lastMissedNotificationCheckTime = userPreference.lastMissedNotificationCheckTime else {
-      return false
-    }
-    let sinceNow = lastMissedNotificationCheckTime.timeIntervalSinceNow
-    // Important: timeIntervalSinceNow is negative if it's in the past
-    return sinceNow < 0 && Int64(abs(sinceNow)) > MISSED_NOTIFICATION_TTL_SEC
-  }
-  
-  private func resetStoredState() {
+
+  /// Remove everything persisted and unschedule all alarms
+  func resetStoredState() {
     TUTSLog("Resetting stored state")
     self.unscheduleAllAlarms(userId: nil)
-    self.userPreference.clear()
-    do {
-      try keychainManager.removePushIdentifierKeys()
-    } catch {
-      TUTSLog("Faied to remove pushIdentifier keys \(error)")
-    }
+    self.alarmPersistor.clear()
   }
   
-  private func rescheduleAlarms() {
+  /// Take the alarms from the persistor and schedule the most recent occurrences
+  func rescheduleAlarms() {
     TUTSLog("Re-scheduling alarms")
-    DispatchQueue.global(qos: .background).async {
-      for notification in self.savedAlarms() {
+    let decryptedAlarms = self.savedAlarms()
+      .compactMap { encryptedAlarm in
         do {
-          try self.scheduleAlarm(notification)
+          return try alarmCryptor.decrypt(alarm: encryptedAlarm)
         } catch {
-          TUTSLog("Error when re-scheduling alarm \(notification) \(error)")
+          TUTSLog("Error when decrypting alarm \(encryptedAlarm) \(error)")
+          return nil
         }
       }
+    let occurences = alarmCalculator.futureOccurrences(acrossAlarms: decryptedAlarms, upToForEach: EVENTS_SCHEDULED_AHEAD, upToOverall: SYSTEM_ALARM_LIMIT)
+    
+    for occurrence in occurences.reversed() {
```

Comment:
using reversed so the closest alarms are scheduled last (since iOS evicts oldest first, i assume?)

Reply:
yes, this comment got lost somewhere